### PR TITLE
feat: Add 4337 hooks

### LINKED ISSUE #23: Enable 4337 configuration
## Context / issue

Currently the `sdk-starter-kit` allows to extend its functionality adding 4337 compatibility. This is not part of the React Hooks yet and it will be a great addition as it will enable sponsoring gas for transactions when using this SDK

## Proposed solution

Enable configuration to allow adding 4337 initialization and extending the sdk-starter-kit.
Create a hook to manage SafeOperations.

It should be possible to:
 - sendSafeOperation
 - confirmSafeOperation
 - getPendingSafeOperations

### PR DESCRIPTION
**Pull request by yagopv**
We are implementing the following updates:

1. A new Context property, `safeOperationOptions`, to pass Safe Operations parameters
2. Hooks for managing Safe Operations:
   - `useSafeOperation`
   - `useSafeOperations`
   - `usePendingSafeOperations`
   - `useSendSafeOperation`
   - `useConfirmSafeOperations`
3. Updating the change configuration hooks (`useUpdateThreshold`, `useUpdateOwners`) to use Safe Operations when the `safeOperationOptions` property is present

### CODE REVIEW

File: src/hooks/useDynamicSafeAction.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  useConfirmTransaction,
+  UseConfirmTransactionReturnType
+} from '@/hooks/useConfirmTransaction.js'
+import { useConfig } from './useConfig.js'
+
+export type UseDynamicSafeActionParams = ConfigParam<SafeConfigWithSigner>
+
+export type UseDynamicSafeActionReturnType = {
+  send: UseMutateFunction<SafeClientResult, Error, any, unknown>
+  sendAsync: UseMutateAsyncFunction<SafeClientResult, Error, any, unknown>
```
Comment (dasanra) [inline src/hooks/useDynamicSafeAction.ts]: Should those `any` be `SendTransactionVariables | SendSafeOperationVariables`

File: src/hooks/useDynamicSafeAction.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  confirm: UseMutateFunction<SafeClientResult, Error, any, unknown>
+  confirmAsync: UseMutateAsyncFunction<SafeClientResult, Error, any, unknown>
```
Comment (dasanra) [inline src/hooks/useDynamicSafeAction.ts]: Should those `any` be `ConfirmTransactionVariables | ConfirmSafeOperationVariables`

File: src/hooks/useDynamicSafeAction.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+  const { confirmTransaction, confirmTransactionAsync }: UseConfirmTransactionReturnType =
+    useConfirmTransaction(params)
+
+  return {
+    send: isSafeOperation ? sendSafeOperation : sendTransaction,
+    sendAsync: isSafeOperation ? sendSafeOperationAsync : sendTransactionAsync,
+    confirm: isSafeOperation ? confirmSafeOperation : confirmTransaction,
+    confirmAsync: isSafeOperation ? confirmSafeOperationAsync : confirmTransactionAsync
+  }
```
Comment (dasanra) [inline src/hooks/useDynamicSafeAction.ts]: What do you think about using fast return and avoiding unnecessay initializations?

```suggestion
  // Use the appropriate hooks based on the presence of safeOperations in the config
  if (isSafeOperation) {
    const { sendSafeOperation, sendSafeOperationAsync }: UseSendSafeOperationReturnType =
      useSendSafeOperation(params)
    const { confirmSafeOperation, confirmSafeOperationAsync }: UseConfirmSafeOperationReturnType =
      useConfirmSafeOperation(params)
    
    return {
      send: sendSafeOperation,
      sendAsync: sendSafeOperationAsync,
      confirm: confirmSafeOperation,
      confirmAsync: confirmSafeOperationAsync
    }
  }

  const { sendTransaction, sendTransactionAsync }: UseSendTransactionReturnType =
    useSendTransaction(params)

  const { confirmTransaction, confirmTransactionAsync }: UseConfirmTransactionReturnType =
    useConfirmTransaction(params)

  return {
    send: sendTransaction,
    sendAsync: sendTransactionAsync,
    confirm: confirmTransaction,
    confirmAsync:confirmTransactionAsync
  }
```

File: src/hooks/useUpdateOwners/useAddOwner.ts [hunk 4]
```diff
@@ -1,8 +1,9 @@
 import { UseMutateAsyncFunction, UseMutateFunction, UseMutationResult } from '@tanstack/react-query'
-import { SafeClient, SafeClientResult } from '@safe-global/sdk-starter-kit'
-import { ConfigParam, SafeConfigWithSigner } from '@/types/index.js'
+import { SafeClientResult } from '@safe-global/sdk-starter-kit'
+import { ConfigParam, SafeConfigWithSigner, SafeClient } from '@/types/index.js'
 import { useSignerClientMutation } from '@/hooks/useSignerClientMutation.js'
-import { useSendTransaction } from '@/hooks/useSendTransaction.js'
+import { useDynamicSafeAction } from '../useDynamicSafeAction.js'
```
Comment (dasanra) [inline src/hooks/useUpdateOwners/useAddOwner.ts]:
```suggestion
import { useDynamicSafeAction } from '@hooks/useDynamicSafeAction.js'
```

File: src/hooks/useUpdateOwners/useRemoveOwner.ts [hunk 5]
```diff
@@ -1,7 +1,7 @@
 import { UseMutateAsyncFunction, UseMutateFunction, UseMutationResult } from '@tanstack/react-query'
-import { SafeClient, SafeClientResult } from '@safe-global/sdk-starter-kit'
-import { ConfigParam, SafeConfigWithSigner } from '@/types/index.js'
-import { useSendTransaction } from '@/hooks/useSendTransaction.js'
+import { SafeClientResult } from '@safe-global/sdk-starter-kit'
+import { ConfigParam, SafeConfigWithSigner, SafeClient } from '@/types/index.js'
+import { useDynamicSafeAction } from '../useDynamicSafeAction.js'
```
Comment (dasanra) [inline src/hooks/useUpdateOwners/useRemoveOwner.ts]:
```suggestion
import { useDynamicSafeAction } from '@hooks/useDynamicSafeAction.js'
```

File: src/hooks/useUpdateOwners/useSwapOwner.ts [hunk 6]
```diff
@@ -1,8 +1,8 @@
 import { UseMutateAsyncFunction, UseMutateFunction, UseMutationResult } from '@tanstack/react-query'
-import { SafeClient, SafeClientResult } from '@safe-global/sdk-starter-kit'
-import { ConfigParam, SafeConfigWithSigner } from '@/types/index.js'
+import { SafeClientResult } from '@safe-global/sdk-starter-kit'
+import { ConfigParam, SafeConfigWithSigner, SafeClient } from '@/types/index.js'
 import { useSignerClientMutation } from '@/hooks/useSignerClientMutation.js'
-import { useSendTransaction } from '@/hooks/useSendTransaction.js'
+import { useDynamicSafeAction } from '../useDynamicSafeAction.js'
```
Comment (dasanra) [inline src/hooks/useUpdateOwners/useSwapOwner.ts]:
```suggestion
import { useDynamicSafeAction } from '@hooks/useDynamicSafeAction.js'
```

File: src/hooks/useDynamicSafeAction.ts (hunk 1 above)
Comment (yagopv) [reply src/hooks/useDynamicSafeAction.ts]: fixed

File: src/hooks/useDynamicSafeAction.ts (hunk 2 above)
Comment (yagopv) [reply src/hooks/useDynamicSafeAction.ts]: fixed

File: src/hooks/useDynamicSafeAction.ts (hunk 3 above)
Comment (yagopv) [reply src/hooks/useDynamicSafeAction.ts]: removed the hook

File: src/hooks/useUpdateOwners/useRemoveOwner.ts (hunk 5 above)
Comment (yagopv) [reply src/hooks/useUpdateOwners/useRemoveOwner.ts]: Unnecessary after the latest changes

File: src/hooks/useUpdateOwners/useSwapOwner.ts (hunk 6 above)
Comment (yagopv) [reply src/hooks/useUpdateOwners/useSwapOwner.ts]: Unnecessary after the latest changes

**Review by dasanra — APPROVED** [review]

**Pull request merged**